### ISSUE TITLE
TS2322 error when using satisfies with strictFunctionTypes

### ISSUE THREAD

**Issue by KuSh** [microsoft/TypeScript#55461]
### 🔎 Search Terms

satisfies strictFunctionTypes 2322

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about strictFunctionTypes

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.3.0-dev.20230821#code/C4TwDgpgBAYgrgOwMYAUBOB7MBnKBeKAbwCgoyoAzDDALim2DQEsEBzAbmIF9jikMEDKACMAhmnxQAFIUrUoXOvGTos2AJT4AfFAAGVWlAAkhA112cKiJMCYCRogF4y5GBUuuqcm2WgjA4NAQ9AzoTMwsFXn5BYEpBSRJyKGAIBgBGOjE0ABpSclSGACYspzyuelFbbAomNKgAJQh+NAATAB4GZjYc6SkwTBw6JpaOrpZWXsQAawQMAHcELU08HXG2dS12IA

### 💻 Code

```ts
type FuncProps = {
    foo: string;
}

const bar = ({ foo }: FuncProps) => `foo: ${foo}`;
function baz({ foo }: FuncProps) { return `foo: ${foo}`; }

const fns = {
    test1: bar,
    test2: baz,
} satisfies Record<string, ((props: Record<string, unknown>) => string)>;
```

### 🙁 Actual behavior

Type '({ foo, bar, baz }: FuncProps) => string' is not assignable to type '(props: Record<string, unknown>) => string'.
  Types of parameters '__0' and 'props' are incompatible.
    Type 'Record<string, unknown>' is missing the following properties from type 'FuncProps': foo, bar, baz

### 🙂 Expected behavior

No error, I'm just using satisfies and type of __0 satisfies the one from props

**Comment by MartinJohns** [comment]
> Type 'Record<string, unknown>' is missing the following properties from type 'FuncProps': foo, bar, baz

I don't see this error in your playground. Please make sure your described behavior matches the code you share.

Your issue can be simplified to:

```typescript
type FuncProps = { foo: string; }

const bar = ({ foo }: FuncProps) => `foo: ${foo}`;

bar satisfies (props: Record<string, unknown>) => string;
```

You're asking "does my value match a function that can accept props with arbitrary amount of properties", but your value does **not** accept props with arbitrary amount of properties. Your props **require** at least the property `foo`, but `Record<string, unknown>` may very well omit this property. I see this working as intended.

You can either make your `FuncProps` an "empty" type by making all properties optional, or you check for `FuncProps` instead of `Record<string, unknown>`.

**Comment by KuSh** [comment]
> I don't see this error in your playground. Please make sure your described behavior matches the code you share.
Strange, I've just opened the playground and got the errors 

<img width="1011" alt="image" src="https://github.com/microsoft/TypeScript/assets/60970/26fe33bc-ce11-4927-8281-ee116af62ee0">

> Your issue can be simplified to:
> 
> ```ts
> type FuncProps = { foo: string; }
> 
> const bar = ({ foo }: FuncProps) => `foo: ${foo}`;
> 
> bar satisfies (props: Record<string, unknown>) => string;
> ```
👍

> You're asking "does my value match a function that can accept props with arbitrary amount of properties", but your value does **not** accept props with arbitrary amount of properties. Your props **require** at least the property `foo`, but `Record<string, unknown>` may very well omit this property. I see this working as intended.
> 
> You can either make your `FuncProps` an "empty" type by making all properties optional, or you check for `FuncProps` instead of `Record<string, unknown>`.
I don't understand "satisfies" like that, especially when I'm able to use it the following way, but your explanation is logical.
```ts
const baz = {
    foo: "test"
} satisfies Record<string, unknown>;
```

**Comment by MartinJohns** [comment]
> Strange, I've just opened the playground and got the errors

The playground error is different. It does not mention "the following properties from type 'FuncProps': foo, bar, baz".

**Comment by KuSh** [comment]
> > Strange, I've just opened the playground and got the errors
> 
> The playground error is different. It does not mention "the following properties from type 'FuncProps': foo, bar, baz".

Oh! I've simplified the test-case AFTER creating the issue and forgot to update the issue. Sorry about that

**Comment by fatcerberus** [comment]
> I don't understand "satisfies" like that, especially when I'm able to use it the following way

Think of it this way: `Dog` is assignable to `Animal` because a dog is an animal, right?  Anything you can do with any animal, you can do with a dog.  But `(x: Dog) => void` is **not** assignable to `(x: Animal) => void` because someone might call the function with a cat, but your implementation can only deal with dogs.

This is called contravariance and is what `strictFunctionTypes` enforces.

**Comment by KuSh** [comment]
> Think of it this way: `Dog` is assignable to `Animal` because a dog is an animal, right? Anything you can do with any animal, you can do with a dog. But `(x: Dog) => void` is **not** assignable to `(x: Animal) => void` because someone might call the function with a cat, but your implementation can only deal with dogs.
> 
> This is called contravariance and is what `strictFunctionTypes` enforces.

There seems (at least to me) to be a contradiction in that case, see for exemple  :

```ts
const baz = { foo: "test" };
baz satisfies Record<string, unknown>; // OK

type Baz = typeof baz;
const bar = (baz: Baz) => `foo: ${baz.foo}`;
bar satisfies (baz: Record<string, unknown>) => string; // Property 'foo' is missing in type 'Record<string, unknown>' but required in type 'Baz'.(1360)
```

So the type of baz satisfies Record<string, unknown> but when I use it as a function parameter it doesn't satisfies it anymore ??

Same thing if I explicitly use a type :
```ts
type Baz = { foo: string; }
const baz: Baz = { foo: "test" };
baz satisfies Record<string, unknown>; // OK

const bar = (baz: Baz) => `foo: ${baz.foo}`;
bar satisfies (baz: Record<string, unknown>) => string; // Property 'foo' is missing in type 'Record<string, unknown>' but required in type 'Baz'.(1360)
```

For me satisfies helps enforce some "lazy" or generic typing without loosing the more specific one implied by the code.
In my exemple it helps enforce baz is at least Record<string, unknown> but keep the more specific type (`{ foo: "string" }`, or if I've used `const baz = { foo: "test" as const }` ; `{ foo: "test" }`
I want to be able to do the same with functions, i.e. make sure my functions are "at least" `(_: Record<string, unknown>) => string` but keep the better, more specific definition and I thought satisfies would allow me to do this and the [documentation](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator) seemed to lean in my direction (but English isn't my mother tongue so I can misunderstand things a bit)

I'll read more on covariance and contravariance thanks for the pointer!

**Comment by fatcerberus** [comment]
`(baz: Record<string, unknown>) => string` is a function which I can legally call, per its signature, as:

```ts
fun({ foo: 42 });
```

`baz` does **not** satisfy it, because it *requires*, per its signature, to be called with a `{ foo: "test" }`.  That's contravariance :)

**Comment by RyanCavanaugh** [comment]
This is a correct error; @fatcerberus 's comment above correctly describes why

**Comment by KuSh** [comment]
So how would you achieve what I'm trying to do ? i.e. ensure function signature respect at least a certain form without compromising final type ?

**Comment by KuSh** [comment]
> But that's not what you're trying to do

I'm pretty sure I know what I want to achieve ^^ I thought `satisfies` would do that but it seems I need something else.

Test case is reduced but in my case I have const structures where I want to enforce global form (i.e. error out on constructions outside of what I allow, specifically functions with multiple args or a first arg that isn't a `Record`) without overriding final type as does `satisfies` when no functions are involved

**Comment by fatcerberus** [comment]
So what you're asking for, if I'm reading this right, is a way to say (using the animal analogy again):
* "I want a function whose parameter is assignable to `Animal`, but the exact type of the parameter doesn't matter."

IMO it doesn't really make sense to specify a type this way, since it doesn't give you enough information to do anything with it.  You couldn't safely pass a random `Animal` to it, since it might require something specific (say, a `Dog`), but you also don't know exactly what kind of animal it wants.  Accordingly, there's no way to specify this type (well... `strictFunctionTypes: false` allows it, but it's unsound).

**Comment by KuSh** [comment]
Using the animal analogy, I'm trying to say :

"I want a variable who satisfies following structure : a Record of functions with a unique parameter who extends Animal"

This will ensure that my variable, which has hundreds of entries, do not get functions that accept multiple parameters or that the first one is anything other than something that extend "Animal"

The trick is that satisfies don't change the final type, it just typecheck it, so the functions will know and tell exactly what specific parameter (say a Dog) they requires, and user of that structure too. But I've ensured that my structure is at least of a certain form.

Again this is what you can achieve with satisfies when you don't use functions :
```ts
type StringRecord = { [k: string]: StringRecord | string }

const records = {
  foo: "foo",
  bar: {
    bar: "bar",
    error: 0, // ts2322
  },
  baz: {
    baz: {
      baz: "baz"
      error: false,// ts2322
    }
    error: null,// ts2322
  }
  error: {},// ts2322
} satisfies StringRecord;
```

Here records final type is the one from its value but we typechecked it against StringRecord

**Comment by fatcerberus** [comment]
> satisfies following structure : a Record of functions with a unique parameter who extends Animal

Yeah, you can't do that with an assignability check (which is all that `satisfies` is), because assignability goes in the opposite direction for function parameters, hence *contra*variance.  afaik tcheck you want to do isn't possible in the current type system.
